src/NumericalAlgorithms/Spectral/ComplexDataView.hpp

```diff
@@ -0,0 +1,174 @@
+///  computations will cause operations which act on real or imaginary parts
+///  individually to have better memory locality (so likely improved cache
+///  performance, but such statements are highly hardware-dependent). However,
+///  this representation will give rise to more copies and allocations.
+///
+/// \note The pair of representations is provided as a means to 'turn a dial' in
+/// optimizations. It is unclear which of these representations will be
+/// preferable, and it may well be the case that different representations are
+/// better for different calculation types or different hardware. Therefore,
+/// when optimizing code which uses libsharp, it is encouraged to profile the
+/// cost of each representation for a computation and choose the one which tests
+/// best.
+enum class ComplexRepresentation { Interleaved, RealsThenImags };
+
+namespace detail {
+//  A storage container for storing sub-vector references ("views") of a
+// ComplexDataVector.
+//
+// Warning: For optimization reasons, mutations applied via member functions or
+// manipulations of data pointed to by pointers returned by member functions
+// *may or may not* be immediately applied to the source vector used to
+// construct the view. Correct use of this class will first perform
+// (potentially several) manipulations of the data via the view, then as a
+// final step flush the data to the source vector via the member function
+// `copy_back_to_source()`. At that point, the vector is guaranteed to be
+// placed in the same state as the view. **The process of constructing a view,
+// then mutating data in both the view and the source vector, then calling
+// `copy_back_to_source()` is considered undefined behavior, and depends on the
+// details of the memory layout chosen.**
+template <ComplexRepresentation Representation>
+class ComplexDataView {
+ public:
+  // The Representation determines the internal data representation
+  static const ComplexRepresentation complex_representation = Representation;
+
+  // The internal storage type
+  using value_type = std::complex<double>;
+
+  // Construct a view which starts at index `offset` of the supplied
+  // vector, and extends for `size` elements
+  ComplexDataView(gsl::not_null<ComplexDataVector*> vector, size_t size,
+                  size_t offset = 0) noexcept;
+
+  // Construct a view which starts at pointer `start` and extends for
+  // `size` elements. Need not be a part of a ComplexDataVector.
+  ComplexDataView(std::complex<double>* start, size_t size) noexcept;
+
+  // For the lifetime of the data view, it points to the same portion of a
+  // single vector, default move assign behavior would contradict that
+  ComplexDataView() = delete;
+  ComplexDataView(ComplexDataView&) = default;
+  ComplexDataView(ComplexDataView&&) = default;
+  ComplexDataView operator=(ComplexDataView&&) = delete;
+  ~ComplexDataView() = default;
+
+  // assign into the view the values of a same-sized ComplexDataVector
```

Comment:
I'll add a comment before these to clarify: this class permits exclusively data assignment. To avoid ambiguous behavior with the two types of view, the reassignment of the pointer to the original data is forbidden.